### PR TITLE
feat: add campaign tracking

### PR DESCRIPTION
**Pull request by kevinpagtakhan**
### Summary

Adds [campaign tracking](https://amplitude.atlassian.net/wiki/spaces/GOV/pages/1795063947/Web+Attribution+Tracking) feature. This includes:

* capturing organic/direct traffic
* campaign based user session
* attribution configs
* session id reconciliation (with better session data management)

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

File: packages/analytics-browser/src/attribution/campaign-parser.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
+import { UTMCookie } from '../storage/utm-cookie';
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/attribution/campaign-parser.ts:1]: dedicated module to parse all campaign related data from url params, cookies, document

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
+import { createIdentifyEvent, Identify } from '@amplitude/analytics-core';
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:1]: dedicated module to emit campaign related data. handles
* persisting campaign data for future comparison
* determining if campaign is new
* converting campaign data to identify event
* tracking event

File: packages/analytics-browser/src/config.ts [hunk 3]
```diff
@@ -54,47 +56,90 @@ export class BrowserConfig extends Config implements IBrowserConfig {
...
     this.partnerId = options?.partnerId;
     this.sessionId = options?.sessionId;
-    this.sessionTimeout = options?.sessionTimeout ?? defaultConfig.sessionTimeout;
     this.trackingOptions = options?.trackingOptions ?? defaultConfig.trackingOptions;
     this.userId = userId;
   }
+
+  get deviceId() {
+    return this.sessionManager.getDeviceId();
+  }
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/config.ts:106]: this is part of better session/cookie storage management. adding getters and setters to allow

```ts
this.config.userId = 'asdf'
```

to also update the cookie storage

this was previously

```ts
this.config.userId = 'asdf'
updateCookies(this.config)
```

which can result to easily forgetting to update cookies

File: packages/analytics-browser/src/plugins/context.ts [hunk 4]
```diff
@@ -31,39 +30,51 @@ export class Context implements BeforePlugin {
...
+    /**
+     * Manages user session triggered by new events
+     */
+    if (!this.isSessionValid()) {
+      // Creates new session
+      this.config.sessionId = Date.now();
+    } // else use previously creates session
+    // Updates last event time to extend time-based session
+    this.config.lastEventTime = Date.now();
+    const time = new Date().getTime();
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/plugins/context.ts:47]: this is the addition in here. updating configs also update cookie storage

File: packages/analytics-browser/src/session-manager.ts [hunk 5]
```diff
@@ -1,31 +1,67 @@
...
-};
-
-export const getCookieName = (apiKey: string, limit = 10) => {
-  return `${AMPLITUDE_PREFIX}_${apiKey.substring(0, limit)}`;
-};
-
-export const getOldCookieName = (apiKey: string) => {
-  return `${AMPLITUDE_PREFIX.toLowerCase()}_${apiKey.substring(0, 6)}`;
-};
+import {
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/session-manager.ts:1]: this module acts as later to connect config and cookie storage with some caching strategy

File: packages/analytics-browser/src/session-manager.ts [hunk 6]
```diff
@@ -1,31 +1,67 @@
...
+
+  setSession(session: Partial<UserSession>) {
+    this.cache = { ...this.cache, ...session };
+    this.storage.set(this.storageKey, this.cache);
+  }
+
+  getSessionId() {
+    // Note: always read directly from storage
+    return this.storage.get(this.storageKey)?.sessionId;
+  }
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/session-manager.ts:30]: this is how session id gets reconciled, sdk will always read from cookie storage for session id. so when cookies are updated by a different tab, other tabs should be able to read new session id as soon as it does.

other user session data are cached to cookies are not accessed and parsed more often then needed

File: packages/analytics-browser/src/attribution/campaign-parser.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  getReferrer(): ReferrerParameters {
+    const data: ReferrerParameters = {
+      referrer: undefined,
+      referring_domain: undefined,
+    };
+    try {
+      data.referrer = document.referrer || undefined;
+      data.referring_domain = data.referrer?.split('/')[2] ?? undefined;
+    } catch {
+      //
```
Comment (haoliu-amp) [inline packages/analytics-browser/src/attribution/campaign-parser.ts:64]: If you wanna no ops, you can put it into the comment.

**Review by qingzhuozhen — APPROVED** [review]
lgtm overall, ideally this can be break down to smaller PRs for easier reviews

**Review by justin-fiedler — APPROVED** [review]

File: packages/analytics-browser/src/attribution/campaign-parser.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  UTM_CAMPAIGN,
+  UTM_CONTENT,
+  UTM_MEDIUM,
+  UTM_SOURCE,
+  UTM_TERM,
+  UTMZ_SOURCE,
+  UTMZ_MEDIUM,
+  UTMZ_CAMPAIGN,
+  UTMZ_TERM,
+  UTMZ_CONTENT,
```
Comment (kickerbear) [inline packages/analytics-browser/src/attribution/campaign-parser.ts:13]: Is UTMZ used by any customers?

Comment (kickerbear) [reply packages/analytics-browser/src/attribution/campaign-parser.ts:13]: nvm, see they're cookies

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    this.tracker = options.tracker ?? (() => Promise.resolve());
+    this.excludeReferrers = options.excludeReferrers ?? [];
+    if (typeof location !== undefined) {
+      this.excludeReferrers.unshift(location.hostname);
+    }
+    this.initialEmptyValue = options.initialEmptyValue ?? EMPTY_VALUE;
+    this.onNewCampaign = options.onNewCampaign;
+  }
+
+  isNewCampaign(currentCampaign: Campaign, previousCampaign?: Campaign) {
```
Comment (kickerbear) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:34]: This is the logic I think we'll want to be able to parameterize maybe in a second iteration. Specifically, what if they just wanted to reset on new referrer but not new campaign Id?

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    const isReferrerExcluded =
+      currentCampaign.referring_domain && this.excludeReferrers.includes(currentCampaign.referring_domain);
+    const hasCampaign = Boolean(currentCampaign.utm_campaign);
+    const isNewCampaign = previousCampaign?.utm_campaign !== currentCampaign.utm_campaign;
```
Comment (kickerbear) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:38]: Super tiny Nit you can choose to ignore: Consider using a different variable name given this is also the name of the function and makes searching just a little bit harder

File: packages/analytics-browser/src/browser-client.ts [hunk 11]
```diff
@@ -26,18 +27,32 @@ export class AmplitudeBrowser extends AmplitudeCore<BrowserConfig> {
...
     // Step 4: Track attributions
-    void this.trackAttributions();
+    if (!options?.attribution?.disabled) {
+      await this.trackCampaign(options?.attribution?.excludeReferrer, options?.attribution?.initialEmptyValue);
+    }
+  }
+
+  async trackCampaign(excludeReferrers?: string[], initialEmptyValue?: string) {
+    const tracker = this.track.bind(this);
+    const onNewCampaign = this.setSessionId.bind(this, Date.now());
```
Comment (kickerbear) [inline packages/analytics-browser/src/browser-client.ts:46]: What is this line doing? It looks like it's setting the onNewCampaign which I thought was a boolean to the result of binding the session id to right now...

Comment (kickerbear) [reply packages/analytics-browser/src/browser-client.ts:46]: I thought this behavior was going to be controlled by session definition, which is why I ask - is that somehow achieved by this line?

**Review by kickerbear — COMMENTED** [review]
(In progress review) - Read through campaign tracking and looks good

File: packages/analytics-browser/src/attribution/campaign-tracker.ts (hunk 10 above)
Comment (kevinpagtakhan) [reply packages/analytics-browser/src/attribution/campaign-tracker.ts:38]: ah yes, will think of something different!

File: packages/analytics-browser/src/browser-client.ts (hunk 11 above)
Comment (kevinpagtakhan) [reply packages/analytics-browser/src/browser-client.ts:46]: this creates a callback function that the "attribution module" will call if a new campaign is found. intention is to keep "attribution module" separate from this amplitude client; allowing to be reused with other amplitude client versions like amplitude-js, if we go that route.

**Review by dnshi — APPROVED** [review]

File: packages/analytics-browser/src/browser-client.ts [hunk 12]
```diff
@@ -26,18 +27,32 @@ export class AmplitudeBrowser extends AmplitudeCore<BrowserConfig> {
...
-    updateCookies(this.config, oldCookies.lastEventTime);
-
-    // Step 4: Install plugins
+    // Step 3: Install plugins
     await this.add(new Context());
     await this.add(new Destination());
 
     // Step 4: Track attributions
-    void this.trackAttributions();
+    if (!options?.attribution?.disabled) {
```
Comment (kickerbear) [inline packages/analytics-browser/src/browser-client.ts:39]: Where is the old behavior preserved? Did we say if this is disabled, it keeps the old behavior?

Comment (kevinpagtakhan) [reply packages/analytics-browser/src/browser-client.ts]: Old behavior does not exist in this package. Only new or nothing.

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+    const isReferrerExcluded = Boolean(
+      currentCampaign.referring_domain && this.excludeReferrers.includes(currentCampaign.referring_domain),
+    );
+    const hasNewUtm =
+      previousCampaign.utm_campaign !== currentCampaign.utm_campaign ||
+      previousCampaign.utm_source !== currentCampaign.utm_source ||
+      previousCampaign.utm_medium !== currentCampaign.utm_medium ||
+      previousCampaign.utm_term !== currentCampaign.utm_term ||
+      previousCampaign.utm_content !== currentCampaign.utm_content;
+    const hasNewReferrer = previousCampaign.referring_domain !== currentCampaign.referring_domain;
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:53]: if any utm param or referrer changes, it is a new campaign

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+    const currentCampaign = this.parser.parse();
+    const previousCampaign = this.getCampaignFromStorage();
+    if (!force) {
+      if (!this.trackNewCampaigns || !this.isNewCampaign(currentCampaign, previousCampaign)) {
+        return;
+      }
+      this.onNewCampaign(currentCampaign);
+    }
+    await this.track(this.createCampaignEvent(currentCampaign));
+    this.saveCampaignToStorage(currentCampaign);
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:108]: updated logic of  whether to track or not. `force` parameter is passed with `isNewSession` from client

File: packages/analytics-browser/src/attribution/campaign-tracker.ts [hunk 15]
```diff
@@ -33,6 +34,7 @@ export class CampaignTracker implements ICampaignTracker {
 
     this.disabled = Boolean(options.disabled);
     this.trackNewCampaigns = Boolean(options.trackNewCampaigns);
+    this.trackPageViews = options.trackPageViews === undefined || Boolean(options.trackPageViews);
```
Comment (kevinpagtakhan) [inline packages/analytics-browser/src/attribution/campaign-tracker.ts:37]: optional page view tracking

Comment (kevinpagtakhan) [reply packages/analytics-browser/src/attribution/campaign-tracker.ts:37]: @kickerbear

**Review by kickerbear — APPROVED** [review]
LGTM! I walked through this with Kevin 1:1 as he was building this. Approved

**Pull request merged**